File: generator_web_app/src/modules/api/projects.ts
// API methods for managing projects
// Author: Michael B. Lance

//---------------------------------------------------------------------------------------------------------------------------//

import { Project, Model, HerdUnit } from '@/types/generatorobjects.ts';
import type { ProjectIntf, ModelIntf , HerdUnitIntf} from '@/types/generatorobjects.ts';
import { ApiError } from '@/modules/api/errors.ts'
import { api_url } from '@/modules/api/apiV1Methods.ts';

//---------------------------------------------------------------------------------------------------------------------------//

export async function getProjectModels(project_id: string): Promise<Model[] | undefined> {
    const response = await fetch(`${api_url}/projects/${project_id}/models`, {
        method: 'GET',
        credentials: 'include',
        headers: {
            'Content-Type': 'application/json',
        },
    });
    if (!response.ok) throw new ApiError(await response.json());

    const resp = await response.json();
    let models = [];
    for (const model of resp) models.push(new Model(model as ModelIntf));

    return models;
}

//---------------------------------------------------------------------------------------------------------------------------//

export async function getProjectHerdUnits(project_id: string): Promise<HerdUnit[]> {
    const response = await fetch(`${api_url}/projects/${project_id}/herd-units`, {
        method: 'GET',
        credentials: 'include',
        headers: {
            'Content-Type': 'application/json',
        },
    });
    if (!response.ok) throw new ApiError(await response.json());

    const resp = await response.json();
    let herd_units = [];
    for (const herd_unit of resp) herd_units.push(new HerdUnit(herd_unit as HerdUnitIntf));
    
    return herd_units;
}

File: annotation_api/app/routers/projects/projects.py
# Endpoints for managing projects in the API 
# Author: Michael B. Lance

#---------------------------------------------------------------------------------------------------------------------------#

from uuid import UUID

from flask import Blueprint, abort
from flask_login import login_required
from flask_pydantic import validate
from psycopg.errors import DatabaseError

from app.extensions import base, s3
from database import ObjectNotFound

from .project_validators import *

projectBp = Blueprint('projects', __name__, url_prefix='/api/v1/projects')

#---------------------------------------------------------------------------------------------------------------------------#
# GET

@projectBp.get('/<string:project_id>')
@login_required
def get_by_id(project_id: str):
	'''
	Request a project object from the database using its UUID.
	---
	parameters:
	  - project_id
		in: path
		type: string
		required: true
	responses:
	  200:
		description: The requested project was found.
	  400:
		description: Invalid UUID format provided.
	  404:
		description: No project record found for the provided ID.
	  500:
		description: Database Error.
	'''
	try:
		project = base.get_project(UUID(project_id))
	except ValueError as e:
		abort(400, str(e))
	except ObjectNotFound as e:
		abort(404, str(e))
	except (DatabaseError, Exception) as e:
		abort(500)
	
	return project.to_dict(), 200

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

@projectBp.get('/<string:project_id>/models')
@login_required
def get_models(project_id: str):
	'''
	Retrieve models for a specific project
	---
	---
	parameters:
		- name: project_id
		in: path
		type: string
		required: true
	responses:
	  200:
		description: List of models.
	  404:
		description: No project found.
	  500:
		description: Database error.
	'''
	try:
		models = base.get_project_models(UUID(project_id))
	except ValueError as e:
		abort(400, str(e))
	except ObjectNotFound as e:
		abort(404, str(e))
	except (DatabaseError, Exception):
		abort(500)
	
	return [model.to_dict() for model in models], 200

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

@projectBp.get('/<string:project_id>/herd-units')
@login_required
def get_herd_units(project_id: str):
	'''
	Retrieve herd units for a specific project
	---
	responses:
	  200:
		description: List of herd units.
	  404:
		description: No project found.
	  500:
		description: Database error.
	'''
	try:
		herd_units = base.get_project_herd_units(UUID(project_id))
	except ValueError as e:
		abort(400, str(e))
	except ObjectNotFound as e:
		abort(404, str(e))
	except (DatabaseError, Exception) as e:
		print(e)
		abort(500)
	
	return [herd_unit.to_dict() for herd_unit in herd_units], 200